Tests/AsyncHTTPClientTests/HTTPRequestStateMachineTests.swift

```diff
@@ -0,0 +1,137 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2021 Apple Inc. and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+@testable import AsyncHTTPClient
+import NIO
+import NIOHTTP1
+import XCTest
+
+class HTTPRequestStateMachineTests: XCTestCase {
+    func testSimpleGETRequest() {
+        var state = HTTPRequestStateMachine(isChannelWritable: true, idleReadTimeout: nil)
+        XCTAssertEqual(state.start(), .verifyRequest)
+        let requestHead = HTTPRequestHead(version: .http1_1, method: .GET, uri: "/")
+        XCTAssertEqual(state.requestVerified(requestHead), .sendRequestHead(requestHead, startBody: false, startReadTimeoutTimer: nil))
+
+        let responseHead = HTTPResponseHead(version: .http1_1, status: .ok)
+        XCTAssertEqual(state.receivedHTTPResponseHead(responseHead), .forwardResponseHead(responseHead))
+        let responseBody = ByteBuffer(bytes: [1, 2, 3, 4])
+        XCTAssertEqual(state.receivedHTTPResponseBodyPart(responseBody), .forwardResponseBodyPart(responseBody, resetReadTimeoutTimer: nil))
+        XCTAssertEqual(state.receivedHTTPResponseEnd(), .forwardResponseEnd(readPending: false, clearReadTimeoutTimer: false))
+    }
+
+    func testPOSTRequestWithWriterBackpressure() {
+        var state = HTTPRequestStateMachine(isChannelWritable: true, idleReadTimeout: nil)
+        XCTAssertEqual(state.start(), .verifyRequest)
+        let requestHead = HTTPRequestHead(version: .http1_1, method: .POST, uri: "/", headers: HTTPHeaders([("content-length", "4")]))
+        XCTAssertEqual(state.requestVerified(requestHead), .sendRequestHead(requestHead, startBody: true, startReadTimeoutTimer: nil))
+        let part0 = IOData.byteBuffer(ByteBuffer(bytes: [0]))
+        let part1 = IOData.byteBuffer(ByteBuffer(bytes: [1]))
+        let part2 = IOData.byteBuffer(ByteBuffer(bytes: [2]))
+        let part3 = IOData.byteBuffer(ByteBuffer(bytes: [3]))
+        XCTAssertEqual(state.requestStreamPartReceived(part0), .sendBodyPart(part0))
+        XCTAssertEqual(state.requestStreamPartReceived(part1), .sendBodyPart(part1))
+
+        // oh the channel reports... we should slow down producing...
+        XCTAssertEqual(state.writabilityChanged(writable: false), .pauseRequestBodyStream)
+
+        // but we issued a .produceMoreRequestBodyData before... Thus, we must accept more produced
```

Comment:
Yes, this might happen because of race conditions, if we receive `channelNotWritable` in thread A and we need to communicate it to thread B, B may have already created new data. For this reason, we need to accept more data even though we asked the producer to pause.